marconi-chain-index/src/Marconi/ChainIndex/Experimental/Api/HttpServer.hs

```diff
@@ -0,0 +1,339 @@
+  , _configParams :: !Aeson.Value
+  , _configQueryables :: !MarconiChainIndexQueryables
+  }
+
+makeLenses 'HttpServerConfig
+
+mkRequestTracer :: (MonadIO m) => Trace IO Text -> m Middleware
+mkRequestTracer trace =
+  liftIO $
+    mkRequestLogger
+      def
+        { destination = Callback $ logDebug trace . Text.decodeUtf8 . fromLogStr
+        , outputFormat = DetailedWithSettings def{mPrelogRequests = True}
+        }
+
+-- | Bootstraps the HTTP server
+runHttpServer :: ReaderT HttpServerConfig IO ()
+runHttpServer = do
+  config <- ask
+  let settings = setPort (config ^. configPort) defaultSettings
+  requestTracer <- mkRequestTracer (config ^. configTrace)
+  lift $ runSettings settings $ requestTracer $ marconiApp config
+
+marconiApp :: HttpServerConfig -> Application
+marconiApp config =
+  serve (Proxy @API) $
+    hoistServer (Proxy @API) (hoistHandler config) httpRpcServer
+
+type ReaderHandler env = ExceptT ServerError (ReaderT env IO)
+type ReaderServer env api = ServerT api (ReaderHandler env)
+
+hoistHandler :: HttpServerConfig -> ReaderHandler HttpServerConfig a -> Handler a
+hoistHandler config = Handler . ExceptT . catchExceptions (config ^. configTrace) . flip runReaderT config . runExceptT
+
+catchExceptions :: Trace IO Text -> IO (Either ServerError a) -> IO (Either ServerError a)
+catchExceptions trace action =
+  action
+    `catches` [ Exception.Handler $ \(e :: ServerError) -> pure $ Left e
+              , Exception.Handler $ \(e :: SomeAsyncException) -> throwIO e
+              , Exception.Handler $ \(e :: SomeException) -> do
+                  logError trace . Text.pack $ "Exception caught while handling HTTP request: " <> displayException e
+                  let
+                    -- TODO: Provide better info, especially the request ID
+                    jsonErr = Errors @Int @Aeson.Value Nothing $ mkJsonRpcInternalErr Nothing
+                  pure . Left $
+                    err500
+                      { errHeaders = [(hContentType, "application/json; charset=utf-8")]
+                      , errReasonPhrase = "Internal server error"
+                      , errBody = Aeson.encode jsonErr
+                      }
+              ]
+
+jsonRpcServer :: ReaderServer HttpServerConfig JsonRpcAPI
+jsonRpcServer =
+  echo
+    -- :<|> getTargetAddressesQueryHandler
+    -- :<|> getCurrentSyncedBlockHandler
+    -- :<|> getAddressUtxoHandler
+    -- :<|> getMintingPolicyHashTxHandler
+    -- :<|> getEpochStakePoolDelegationHandler
+    :<|> getEpochNonceHandler
+
+restApiServer :: ReaderServer HttpServerConfig RestAPI
+restApiServer =
+  getTimeHandler
+    :<|> getParamsHandler
+    :<|> getTargetAddressesHandler
+    :<|> getMetricsHandler
+
+httpRpcServer :: ReaderServer HttpServerConfig API
+httpRpcServer = jsonRpcServer :<|> restApiServer
+
+-- | Convenience function for using @MVar@-based indexers
+viewMVar :: (MonadReader s m, MonadIO m) => Getting (MVar a) s (MVar a) -> m a
```

Comment:
Do you need `MonadReader` here?

Reply:
I think so, yes. The [type](https://hackage.haskell.org/package/lens-5.2.3/docs/Control-Lens-Combinators.html#v:view) of `view` is

```haskell
view :: MonadReader s m => Getting a s a -> m a
```

Reply:
Oh I didn't notice that `view` relied on the `MonadReader`, my bad.